### PR TITLE
Added InMemoryCache parameter to FS structure.

### PR DESCRIPTION
**Pull request by dgrr**
This option allows programmer to handle fs files in memory.
Another two functions was added for this implementation:

- NewBufferReader which allows programmer to obtain reader from content byte slice.
- ReleaseBuffer which allows programmer to return byte buffer to pool.

Also I added a test to fs_test.go

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -402,10 +413,33 @@ type fsFile struct {
...
 
+// NewBufferReader returns io.Reader of cache content of fsFile.
+func (ff *fsFile) NewBufferReader() (io.Reader, error) {
+	if ff.content == nil {
+		return nil, errors.New("this file does not have any content")
+	}
+
+	b := ff.bufferPool.Get()
+	if b == nil {
+		b = bytes.NewBuffer(ff.content)
```
Comment (erikdubbelboer) [inline fs.go]: Since you only need an `io.Reader` using `bytes.NewReader` is much better.

File: fs.go [hunk 2]
```diff
@@ -770,11 +805,22 @@ func (h *fsHandler) handleRequest(ctx *RequestCtx) {
...
 	}
 
-	r, err := ff.NewReader()
-	if err != nil {
-		ctx.Logger().Printf("cannot obtain file reader for path=%q: %s", path, err)
-		ctx.Error("Internal Server Error", StatusInternalServerError)
-		return
+	var r io.Reader
+	if loaded {
+		r, err = ff.NewBufferReader()
```
Comment (erikdubbelboer) [inline fs.go]: When an `Accept-Ranges` header is send, code below will unconditionally casts `r` to an `io.Closer` and `byteRangeUpdater` both of which aren't implemented by `bytes.Buffer` and will cause a panic.

File: fs.go [hunk 3]
```diff
@@ -1158,10 +1204,22 @@ func (h *fsHandler) newFSFile(f *os.File, fileInfo os.FileInfo, compressed bool)
...
 	}
 
+	var buffer []byte
+	if !store {
+		buffer = nil
+	} else {
+		buffer = make([]byte, contentLength)
+		n, err := f.Read(buffer)
+		if err != nil || n != contentLength {
+			buffer = nil
```
Comment (erikdubbelboer) [inline fs.go:1215]: Probably better to return the error here?

File: fs_test.go [hunk 4]
```diff
@@ -61,6 +61,29 @@ func TestNewVHostPathRewriterMaliciousHost(t *testing.T) {
...
+		InMemoryCache: true,
+	}
+	fs.NewRequestHandler()(&ctx)
+
+	if len(ctx.Response.Body()) == 0 {
+		t.Fatal("zero body for existing file")
+	}
+	if s := ctx.Response.Header.Peek("Content-Encoding"); !bytes.Equal(s, []byte("gzip")) {
+		t.Fatal("Uncompressed")
+	}
```
Comment (erikdubbelboer) [inline fs_test.go]: Maybe check the actual result body as well? And check that multiple requests return the same content. And maybe check that it actually uses the in memory cache by changing the file on disk and checking that the old content is returned. And test `Accept-Ranges`.

**Comment by erikdubbelboer** [comment]
Another useful option might be to set a max size for files that should be cached. Or even better would be a callback that can be used to decide which file should and shouldn't be cached. Or will that be too complicated? I'm thinking of a situation where you want to cache a bunch of static files but maybe not some dynamically generated ones. So the callback should get the file name and size to decide.

File: fs.go (hunk 1 above)
Comment (dgrr) [reply fs.go]: Oh, you are right. I did not see NewReader function.

**Comment by dgrr** [comment]
Yes. When I started this branch I thought to set maximum size or similar parameter for large files but probably the programmer would serve all files without distinguishing it. I didn't think to do it in a callback, maybe this will be a good solution. I am going to implement it in the following pull requests.

File: fs.go (hunk 3 above)
Comment (dgrr) [reply fs.go:1215]: I don't want to break the file serving for reading problems. I can return error code in the final return sentence and maybe handle it out of the function. What do you think?

File: fs_test.go (hunk 4 above)
Comment (dgrr) [reply fs_test.go]: I did not push the good test file. This is another test. I will replace it. Sorry.

**Comment by thehowl** [comment]
Can you make some benchmarks using wrk https://github.com/wg/wrk of serving a, say, ~100kb file, on Linux with the following settings?

- InMemoryCache enabled
- InMemoryCache disabled
- InMemoryCache disabled, execute `vmtouch -vt file` before running server

Using https://godoc.org/golang.org/x/exp/mmap might also be interesting, since the Kernel takes care of it.

But in any case, the reason for this is that I'd like to see whether this brings some real tangible performance improvement or we can just offload the job of caching to memory to Linux.

**Comment by illotum** [comment]
I would recommend https://github.com/giltene/wrk2 instead. 

If the measurements are at low milliseconds `wrk` becomes unreliable.

**Comment by dgrr** [comment]
@erikdubbelboer I consider this pull request is ready for production. I tested in a real case and it works successfully.

**Comment by thehowl** [comment]
The question I asked is not whether it works or not - but rather whether it really brings proper performance improvements. The API of fasthttp is already very big, if instead of expanding it we can give other solutions to caching files in-memory then that would be great IMO

File: fs.go (hunk 2 above)
Comment (erikdubbelboer) [reply fs.go]: @themester you didn't do anything about this?

File: fs.go [hunk 5]
```diff
@@ -770,11 +806,22 @@ func (h *fsHandler) handleRequest(ctx *RequestCtx) {
...
-	r, err := ff.NewReader()
-	if err != nil {
-		ctx.Logger().Printf("cannot obtain file reader for path=%q: %s", path, err)
-		ctx.Error("Internal Server Error", StatusInternalServerError)
-		return
+	var r io.Reader
+	if cache {
+		r, err = ff.NewBufferReader()
+		if err != nil {
+			ctx.Logger().Printf("error reading buffer from %q: %s", ff.f.Name(), err)
```
Comment (erikdubbelboer) [inline fs.go:813]: Having and empty file and caching on will now always print this error. Having an empty file should work normally.

**Comment by erikdubbelboer** [comment]
Just posting this here as well in case you don't notice the outdated comments:
When an `Accept-Ranges` header is send, the code will unconditionally casts `r` to an `io.Closer` and `byteRangeUpdater` both of which aren't implemented by `bytes.Reader` and will cause a panic.

Please add a test with `Accept-Ranges`.

Also some benchmarks would indeed be nice.

File: fs.go (hunk 2 above)
Comment (dgrr) [reply fs.go]: Yes. I changed it here https://github.com/erikdubbelboer/fasthttp/pull/32/commits/ad26055ae8eac5d6f284794b594328e74b5974c8

**Review by dgrr — COMMENTED** [review]
@erikdubbelboer yes, I applied the changes in this commit https://github.com/erikdubbelboer/fasthttp/commit/ad26055ae8eac5d6f284794b594328e74b5974c8
This is my commit history in this pull request. https://github.com/themester/fasthttp/commits/cache
Commits in pull requests mode are very confusing.

I just have to add benchmark about InMemoryCache.

**Comment by erikdubbelboer** [comment]
Sorry I should have been  more clear. You need to test these cases https://github.com/themester/fasthttp/blob/d8baa6e38008bb508c82682753cba376fb93b81d/fs.go#L839 and https://github.com/themester/fasthttp/blob/d8baa6e38008bb508c82682753cba376fb93b81d/fs.go#L845 where your changes will cause panics. As you can see you need to send `Accept-Ranges: bytes` and `Range: bytes=1-2` to trigger one case, and `Range: bytes=1-invalid` to trigger the other. And then you of course also need to test that the body you are getting back with this `Range` header is correct.

**Comment by dgrr** [comment]
Sorry for the inconvenience. I don't wanna to support this pull request. After a while I think this is an unneeded pr.

**Pull request closed without merging**